exp/internal/webhooks/ipaddress.go

```diff
@@ -0,0 +1,168 @@
+	oldIP, ok := oldObj.(*expv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", oldObj))
+	}
+	newIP, ok := newObj.(*expv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", newObj))
+	}
+
+	if !reflect.DeepEqual(oldIP.Spec, newIP.Spec) {
+		return field.Forbidden(field.NewPath("spec"), "the spec of IPAddress is immutable")
+	}
+	return nil
+}
+
+// ValidateDelete implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateDelete(ctx context.Context, obj runtime.Object) error {
+	return nil
+}
+
+func (webhook *IPAddress) validate(ctx context.Context, ip *expv1.IPAddress) error {
+	allErrs := field.ErrorList{}
+
+	addr, err := netaddr.ParseIP(ip.Spec.Address)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.address"),
+				ip.Spec.Address,
+				"not a valid IP address",
+			))
+	}
+
+	if ip.Spec.Prefix < 0 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix cannot be negative",
+			))
+	}
+	if addr.Is4() && ip.Spec.Prefix > 32 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix is too large for an IPv4 address",
+			))
+	}
+	if addr.Is6() && ip.Spec.Prefix > 128 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix is too large for an IPv6 address",
+			))
+	}
+
+	_, err = netaddr.ParseIP(ip.Spec.Gateway)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.gateway"),
+				ip.Spec.Gateway,
+				"not a valid IP address",
+			))
+	}
+
+	claim := &expv1.IPAddressClaim{}
+	err = webhook.Client.Get(ctx, types.NamespacedName{Name: ip.Spec.ClaimRef.Name, Namespace: ip.ObjectMeta.Namespace}, claim)
+	if apierrors.IsNotFound(err) {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.claimRef"),
+				ip.Spec.ClaimRef,
+				"the referenced claim could not be found",
+			))
+	} else if err != nil {
+		allErrs = append(allErrs,
+			field.InternalError(field.NewPath("spec.claimRef"), err))
+	}
+
+	if ip.Spec.PoolRef.APIGroup == nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.poolRef.apiGroup"),
+				ip.Spec.PoolRef.APIGroup,
+				"the pool reference needs to contain a group"))
+	}
+
+	if !(ip.Spec.PoolRef.APIGroup != nil && claim.Spec.PoolRef.APIGroup != nil &&
+		*ip.Spec.PoolRef.APIGroup == *claim.Spec.PoolRef.APIGroup &&
+		ip.Spec.PoolRef.Kind == claim.Spec.PoolRef.Kind &&
+		ip.Spec.PoolRef.Name == claim.Spec.PoolRef.Name) {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.poolRef"),
+				ip.Spec.PoolRef,
+				"the referenced pool is different from the pool referenced by the claim this address should fulfill",
+			))
+	}
```

Comment:
Let's only run this part if we didn't get an error in l.134.

Played a bit around with it and if the claimRef is invalid it's a bit confusing to also get this error (given that Claim couldn't have been retrieved)